File: examples/python/metadata_scope_example.py
"""Scoped memory per user: two users share a DB but not their memories.

Run: python examples/python/metadata_scope_example.py
Requires: OPENAI_API_KEY in .env (for the LLM replies; the scope itself is
  database-only and needs no network).
"""

from __future__ import annotations

import os
import tempfile

from dotenv import load_dotenv
from openai import OpenAI

from membrain import MembrainClient

load_dotenv()

DB_DIR = os.path.join(tempfile.gettempdir(), "membrain_scope_example")
INDEXED = ["user_id"]
llm = OpenAI()


def chat_as(user_id: str, message: str) -> str:
    client = MembrainClient(
        config={"storage": {"backend": "memscaledb", "path": DB_DIR,
                            "indexed_metadata_keys": INDEXED}},
        scope={"user_id": user_id},
    )
    try:
        prior = "\n".join(m.content for m in client.search(message, limit=5).memories)
        reply = llm.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user",
                       "content": f"Context:\n{prior}\n\nUser: {message}"}],
        ).choices[0].message.content
        client.store_observation(f"{user_id}: {message}")
        client.store_observation(f"assistant to {user_id}: {reply}")
        return reply
    finally:
        client.close()


if __name__ == "__main__":
    print(chat_as("alice", "I love rust programming"))
    print(chat_as("bob", "I love go programming"))
    # alice asks for her own memory; only alice-scoped rows are visible.
    print(chat_as("alice", "what do I like?"))

File: examples/javascript/metadata_scope_example.mjs
// Scoped memory per user: two users share a DB but not their memories.
// Run: node examples/javascript/metadata_scope_example.mjs
// Requires: OPENAI_API_KEY for the LLM replies.

import { tmpdir } from "node:os";
import { join } from "node:path";

import OpenAI from "openai";
import { MembrainClient } from "membrain";

const DB_DIR = join(tmpdir(), "membrain_scope_example");
const INDEXED = ["user_id"];
const llm = new OpenAI();

async function chatAs(userId, message) {
  const client = new MembrainClient(
    {
      storage: {
        backend: "memscaledb",
        path: DB_DIR,
        indexed_metadata_keys: INDEXED,
      },
    },
    { scope: { user_id: userId } },
  );
  try {
    const prior = client
      .search(message, 5)
      .memories.map((m) => m.content)
      .join("\n");
    const reply = (
      await llm.chat.completions.create({
        model: "gpt-4o-mini",
        messages: [
          { role: "user", content: `Context:\n${prior}\n\nUser: ${message}` },
        ],
      })
    ).choices[0].message.content;
    client.storeObservation(`${userId}: ${message}`);
    client.storeObservation(`assistant to ${userId}: ${reply}`);
    return reply;
  } finally {
    client.close();
  }
}

console.log(await chatAs("alice", "I love rust programming"));
console.log(await chatAs("bob", "I love go programming"));
console.log(await chatAs("alice", "what do I like?"));
